src/verbs/queue_pair.rs

```diff
@@ -411,11 +431,30 @@ impl<'res> QueuePairBuilder<'res> {
     }
 
     // build extended qp
-    pub fn build_ex(&mut self) -> Result<ExtendedQueuePair<'res>, String> {
-        let qp = unsafe { ibv_create_qp_ex((*self.init_attr.pd).context, &mut self.init_attr).unwrap_or(null_mut()) };
+    pub fn build_ex(&self) -> Result<ExtendedQueuePair<'res>, String> {
+        let mut attr = self.init_attr.clone();
+
+        // to build a real extended qp instead of a basic qp, we need to pass in
+        // these essential attributes.
+        attr.comp_mask |=
+            ibv_qp_init_attr_mask::IBV_QP_INIT_ATTR_PD.0 | ibv_qp_init_attr_mask::IBV_QP_INIT_ATTR_SEND_OPS_FLAGS.0;
+
+        // unless user specified, we assume every extended qp would support send,
```

Comment:
Maybe we can put the default values for `comp_mask` and `send_ops_flags` directly in the `QueuePairBuilder::new()` rather than checking them every time `build_ex` is called.